File: R/samples.R
#' Obtain sample data and info from a ProTracker module
#' 
#' Obtain sample data and info from a ProTracker module at a specific index.
#' ProTracker modules can hold up to 31 samples. The index should range from 0 to 30.
#' @param mod An object of class `pt2mod` from which to obtain sample data and information
#' @param i The index of the requested sample (between 0 and 30).
#' @param ... Ignored.
#' @returns Returns a sample object of class `pt2samp`.
#' @examples
#' mod <- pt2_read_mod(pt2_demo())
#' 
#' smp <- pt2_sample(mod, 0L)
#' @author Pepijn de Vries
#' @include helpers.R
#' @export
pt2_sample <- function(mod, i, ...) {
  .check_mod(mod)
  structure(
    list(mod = mod, i = as.integer(i)),
    class = "pt2samp"
  ) |>
    as.raw()
}

#' @rdname mod_info
#' @include mod_info.R
#' @method pt2_name pt2samp
#' @export
pt2_name.pt2samp <- function(x, ...) {
  attributes(x)$sample_info$text
}

#' @rdname mod_info
#' @include mod_info.R
#' @method pt2_name<- pt2samp
#' @export
`pt2_name<-.pt2samp` <- function(x, ..., value) {
  attr(x, "sample_info")$text <- substr(value, 0, 22)
  validate_sample_raw_(x)
  x
}

#' @rdname mod_info
#' @include mod_info.R
#' @method pt2_name pt2samplist
#' @export
pt2_name.pt2samplist <- function(x, ...) {
  result = list()
  for (i in seq_len(length(x))) {
    result[[i]] <- pt2_name(x[[i]])
  }
  unlist(result)
}

#' @rdname mod_info
#' @include mod_info.R
#' @method pt2_name<- pt2samplist
#' @export
`pt2_name<-.pt2samplist` <- function(x, ..., value) {
  if (length(x) != length(value)) stop("Replacement should have the same length as `x`")
  for (i in seq_len(length(x))) {
    pt2_name(x[[i]]) <- value[i]
  }
  x
}

#' @rdname mod_info
#' @export
pt2_n_sample <- function(mod, ...) {
  count = 0
  for (i in 0L:30L) {
    if (mod_sample_info_(mod, i)$length > 0) count = count + 1
  }
  count
}

#' Get or set ProTracker sample properties
#'
#' Get or set properties of a ProTracker sample. See 'details' section
#' for available properties and associated functions.
#' 
#' ProTracker audio samples hold some meta data that affect their
#' playback. The following functions can be used to get or set these
#' properties:
#' 
#'  * `pt2_finetune()`: When a sample is out of tune, you can use the
#'    use the 'finetune' value to tune the sample. It is an integer value
#'    ranging between -8 and +7.
#'  * `pt2_volume()`: An integer value to adjust the sample volume. It ranges
#'    between 0 (silent) to 64 (maximum volume).
#'  * `pt2_loop_start()` and `pt2_loop_length()`: defines if a sample is looped.
#'    The loop start is defined as an even integer indicating the sample index
#'    (zero based) where the loop should start. The loop length is the number
#'    of samples, over which to loop. The loop length should also be an even
#'    number greater than 0. The sum of the loop start and the loop length
#'    should never be larger than the sample's length.
#'  * `pt2_is_looped()`: a `logical` value, indicating if the sample is looped.
#'    If you set the loop status to `TRUE` when the sample is not looped yet,
#'    the new loop will start at index 0, and will have a length equal to the
#'    sample length.
#' @param sample A module sample of class `pt2samp`.
#' @param ... Ignored
#' @param value Replacement value for the sample property
#' @returns Returns an `integer` or `logical` value. Depending on the
#' called function. See the detail section for specifics.
#' @seealso [pt2_name()]
#' @examples
#' mod <- pt2_read_mod(pt2_demo())
#' 
#' pt2_finetune(mod$samples[[1]])
#' pt2_finetune(mod$samples[[1]]) <- -8L
#' 
#' pt2_volume(mod$samples[[2]])
#' pt2_volume(mod$samples[[2]]) <- 64L
#' 
#' pt2_loop_start(mod$samples[[1]])
#' pt2_loop_start(mod$samples[[1]]) <- 400L
#' 
#' pt2_loop_length(mod$samples[[1]])
#' pt2_loop_length(mod$samples[[1]]) <- 274L
#' 
#' pt2_is_looped(mod$samples[[2]])
#' pt2_is_looped(mod$samples[[2]]) <- TRUE
#' @rdname sample_properties
#' @export
pt2_finetune <- function(sample, ...) {
  return(attr(sample, "sample_info")$fineTune)
}

#' @rdname sample_properties
#' @export
`pt2_finetune<-` <- function(sample, ..., value) {
  attr(sample, "sample_info")$fineTune <- as.integer(value)
  validate_sample_raw_(sample)
  return(sample)
}

#' @rdname sample_properties
#' @export
pt2_volume <- function(sample, ...) {
  return(attr(sample, "sample_info")$volume)
}

#' @rdname sample_properties
#' @export
`pt2_volume<-` <- function(sample, ..., value) {
  attr(sample, "sample_info")$volume <- as.integer(value)
  validate_sample_raw_(sample)
  return(sample)
}

#' @rdname sample_properties
#' @export
pt2_loop_start <- function(sample, ...) {
  return(attr(sample, "sample_info")$loopStart)
}

#' @rdname sample_properties
#' @export
`pt2_loop_start<-` <- function(sample, ..., value) {
  attr(sample, "sample_info")$loopStart <- as.integer(value - (value %% 2))
  validate_sample_raw_(sample)
  return(sample)
}

#' @rdname sample_properties
#' @export
pt2_loop_length <- function(sample, ...) {
  return(attr(sample, "sample_info")$loopLength)
}

#' @rdname sample_properties
#' @export
`pt2_loop_length<-` <- function(sample, ..., value) {
  attr(sample, "sample_info")$loopLength <- as.integer(value - (value %% 2L))
  validate_sample_raw_(sample)
  return(sample)
}

#' @rdname sample_properties
#' @export
pt2_is_looped <- function(sample, ...) {
  si <- attr(sample, "sample_info")
  return(!(si$loopStart == 0L && si$loopLength == 2L))
}

#' @rdname sample_properties
#' @export
`pt2_is_looped<-` <- function(sample, ..., value) {
  if (!is.logical(value) || length(value) != 1)
    stop("Replacement value needs to be a single logical value")
  is_looped <- pt2_is_looped(sample)
  if (is_looped && !value) {
    attr(sample, "sample_info")$loopStart <- 0L
    attr(sample, "sample_info")$loopLength <- 2L
  }
  if (!is_looped && value) {
    attr(sample, "sample_info")$loopStart <- 0L
    attr(sample, "sample_info")$loopLength <- attr(sample, "sample_info")$length
  }
  validate_sample_raw_(sample)
  return(sample)
}

#' Coerce ProTracker sample to audio sample
#'
#' Coerce a sample from a ProTracker module to an `audio` package
#' sample. Note that this function differs from [`pt2_render()`] as
#' it will not mimic Commodore Amiga hardware. It will just
#' pass the pure sample data.
#' @param sample A ProTracker sample of class `pt2samp`.
#' @param note A character representing a note (`"C-3"` by default).
#' It is used to calculate the playback rate for the sample.
#' @param finetune An `integer` fine tune value (between -8 and 7). Used
#' to tune the sample's note.
#' @param loop An `integer` value indicating how often the sample
#' should be looped (if it is looped). It will be ignored when `loop` has
#' a value of zero (or less), or when the sample is not looped
#' (see [`pt2_is_looped()`]).
#' @param options Options used for calculating play back rate. See
#' [`pt2_render_options()`] for all available options. But note that
#' not all options affect the play back rate.
#' @param ... Ignored
#' @returns Returns an audio sample of class [`audio::audioSample()`].
#' @examples
#' mod <- pt2_read_mod(pt2_demo())
#' aud <- pt2_sample_to_audio(mod$samples[[1]])
#' @export
pt2_sample_to_audio <- function(
    sample, note = "C-3", finetune = 0, options = pt2_render_options(),
    loop = 20L, ...) {
  per <- pt2_note_to_period(note, "-", finetune)
  sample_freq <- pt_get_PAL_hz() / per
  sample_info <- attributes(sample)$sample_info
  is_looped   <- pt2_is_looped(sample)
  sample <- -1 + (as.integer(sample) + 128L) / 127.5
  attributes(sample) <- NULL
  idx <- length(sample)
  if (is_looped && loop > 0L) {
    loop_idx <- with(sample_info, loopStart + seq_len(loopLength))
    loop_idx <- rep(loop_idx, times = loop)
    idx <- c(seq_len(sample_info$loopStart),
             loop_idx)
  }
  sample <- sample[idx]
  audio::audioSample(sample, rate = sample_freq)
}

File: src/samples.cpp
#include <cpp11.hpp>
#include "get_mod.h"
#include "pt2-clone.h"
using namespace cpp11;

moduleSample_t * get_mod_sampinf_internal(module_t * my_song, int idx) {
  if (idx < 0 || idx >= MOD_SAMPLES) stop("Index out of range");
  
  return &my_song->samples[idx];
}

list mod_sample_info_internal2(moduleSample_t * samp) {

  writable::list attr({
    "length"_nm       = (int)samp->length,
      "loopStart"_nm  = (int)samp->loopStart,
      "loopLength"_nm = (int)samp->loopLength,
      "fineTune"_nm   = (int)samp->fineTune,
      "volume"_nm     = (int)samp->volume,
      "text"_nm       = r_string(samp->text)
  });
  return attr;
}

list mod_sample_info_internal(module_t * my_song, int idx) {
  moduleSample_t * samp = get_mod_sampinf_internal(my_song, idx);
  return mod_sample_info_internal2(samp);
}

raws mod_sample_as_raw_internal(module_t * my_song, int idx) {
  moduleSample_t * samp = get_mod_sampinf_internal(my_song, idx);
  int8_t *sampleData = &my_song->sampleData[samp->offset];
  uint32_t len = samp->length;
  writable::raws sampledata((R_xlen_t)len);
  uint8_t * sampdest = (uint8_t *)RAW(as_sexp(sampledata));
  
  memcpy(sampdest, sampleData, len);

  SEXP attr = mod_sample_info_internal(my_song, idx);
  sampledata.attr("class") = "pt2samp";
  sampledata.attr("sample_info") = attr;
  return sampledata;
}

[[cpp11::register]]
raws mod_sample_as_raw_(SEXP mod, int idx) {
  module_t *my_song = get_mod(mod);
  return mod_sample_as_raw_internal(my_song, idx);
}

integers mod_sample_as_int_internal(module_t * my_song, int idx) {
  moduleSample_t * samp = get_mod_sampinf_internal(my_song, idx);
  int8_t *sampleData = &my_song->sampleData[samp->offset];
  uint32_t len = samp->length;
  writable::integers sampledata((R_xlen_t)len);
  for (uint32_t j = 0; j < len; j++) {
    sampledata.at((int)j) = (int)sampleData[j];
  }
  
  SEXP attr = mod_sample_info_internal(my_song, idx);
  sampledata.attr("sample_info") = attr;
  return sampledata;
}

[[cpp11::register]]
list mod_sample_info_(SEXP mod, int idx) {
  module_t *my_song = get_mod(mod);
  return mod_sample_info_internal(my_song, idx);
}

[[cpp11::register]]
integers mod_sample_as_int_(SEXP mod, int idx) {
  module_t *my_song = get_mod(mod);
  return mod_sample_as_int_internal(my_song, idx);
}

[[cpp11::register]]
logicals validate_sample_raw_(raws smp_data) {
  bool result = true;
  list my_attr   = list(smp_data.attr("sample_info"));
  int length     = writable::integers(my_attr["length"]).at(0);
  int loopStart  = writable::integers(my_attr["loopStart"]).at(0);
  int loopLength = writable::integers(my_attr["loopLength"]).at(0);
  int volume     = writable::integers(my_attr["volume"]).at(0);
  int fineTune   = writable::integers(my_attr["fineTune"]).at(0);
  r_string text  = writable::strings(my_attr["text"]).at(0);

  // Sample size should be even
  result = result &&
    !(length % 2 == 1 || smp_data.size() != length || length < 0 || length > config.maxSampleLength);

  result = result &&
    !(loopStart % 2 == 1 || loopStart < 0 || loopStart > length);

  result = result &&
    !(loopLength % 2 == 1 || loopLength < 2 || (loopStart + loopLength) > length);

  result = result &&
    !(fineTune < 0 || fineTune > 0xf);

  result = result &&
    !(volume < 0 || volume > 64);

  result = result &&
    !(text.size() > 22);

  writable::logicals result_sexp((R_xlen_t)1);
  result_sexp.at(0) = result;

  return result_sexp;
}

[[cpp11::register]]
SEXP mod_set_sample_(SEXP mod, int idx, raws smp_data) {
  module_t *my_song = get_mod(mod);
  moduleSample_t * samp = get_mod_sampinf_internal(my_song, idx);
  int8_t *sampleData = &my_song->sampleData[samp->offset];
  // check validity of smp_data
  validate_sample_raw_(smp_data);
  
  list my_attr = list(smp_data.attr("sample_info"));
  
  memset(&samp->text, 0, 22);
  
  if (my_attr == R_NilValue) {
    samp->length     = smp_data.size();
    samp->loopStart  = 0;
    samp->loopLength = 2;
    samp->fineTune   = 0;
    samp->volume     = 64;
  } else {
    samp->length     = writable::integers(my_attr["length"]).at(0);
    samp->loopStart  = writable::integers(my_attr["loopStart"]).at(0);
    samp->loopLength = writable::integers(my_attr["loopLength"]).at(0);
    samp->fineTune   = writable::integers(my_attr["fineTune"]).at(0);
    samp->volume     = writable::integers(my_attr["volume"]).at(0);
    r_string name    = writable::strings(my_attr["text"]).at(0);
    int len = name.size();
    len = len > 22 ? 22 : len;
    memcpy(&samp->text, ((std::string)name).c_str(), len);
  }
  uint8_t * buffer = (uint8_t *)RAW(as_sexp(smp_data));
  memcpy(sampleData, buffer, smp_data.size());
  
  return mod;
}
